src/main/java/org/radarcns/hdfs/util/PostponedWriter.java

```diff
@@ -0,0 +1,117 @@
+package org.radarcns.hdfs.util;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.Flushable;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * File writer where data is written in a separate thread with a timeout.
+ */
+public abstract class PostponedWriter implements Closeable, Flushable {
+    private static final Logger logger = LoggerFactory.getLogger(PostponedWriter.class);
+
+    private final ScheduledExecutorService executor;
+    private final long timeout;
+    private final TimeUnit timeoutUnit;
+    private final String name;
+    private final AtomicReference<Future<?>> writeFuture;
+    private Path tempDir;
+
+    /**
+     * Constructor with timeout.
+     * @param name thread name.
+     * @param timeout maximum time between triggering a write and actually writing the file.
+     * @param timeoutUnit timeout unit.
+     */
+    public PostponedWriter(String name, long timeout, TimeUnit timeoutUnit) {
+        this.name = name;
+        executor = Executors.newScheduledThreadPool(1, r -> new Thread(r, name));
+        this.timeout = timeout;
+        this.timeoutUnit = timeoutUnit;
+        writeFuture = new AtomicReference<>(null);
+    }
+
+    /**
+     * Trigger a write to occur within set timeout. If a write was already triggered earlier but has
+     * not yet taken place, the write will occur earlier.
+     */
+    public void triggerWrite() {
+        Future<?> localWriteFuture = writeFuture.get();
+        if (localWriteFuture == null) {
+            localWriteFuture = executor.schedule(this::startWrite, timeout, timeoutUnit);
+            if (!writeFuture.compareAndSet(null, localWriteFuture)) {
+                localWriteFuture.cancel(false);
+            }
+        }
+    }
+
+    /** Start the write in the writer thread. */
+    protected void startWrite() {
+        writeFuture.set(null);
+        doWrite();
+    }
+
+    /** Perform the write. */
+    protected abstract void doWrite();
+
+    @Override
+    public void close() throws IOException {
+        doFlush(true);
+        try {
+            executor.awaitTermination(5, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            logger.error("Failed to write data {}", name, e);
```

Comment:
Should be one more `{}` here otherwise it just says - `Failed to write data: timeout`

Reply:
The "timeout" message comes from a few lines lower. I've increased the timeouts to 30 seconds now.